devito/stencilkernel.py

```diff
@@ -29,11 +39,22 @@ def __init__(self, stencils, name="Kernel", compiler=None):
         stencils = stencils if isinstance(stencils, list) else [stencils]
         self.expressions = [Expression(s) for s in stencils]
 
+        # Lower all expressions to "indexed" API
+        for e in self.expressions:
+            e.indexify()
+
+        # Apply supplied substitutions
+        if subs is not None:
+            for expr in self.expressions:
+                expr.substitute(subs)
```

Comment:
Sorry, I completely misunderstood what's going on here. That's OK, forget about my comments (apart from the 2n vs n^2 which you may clarify?)